File: regression/hooks/runF77/hookstest.f
c
c     File:       hookstest.f
c     Copyright:  (c) 2001 Lawrence Livermore National Security, LLC
c     Revision:   @(#) $Revision: 5006 $
c     Date:       $Date: 2005-11-01 15:20:38 -0800 (Tue, 01 Nov 2005) $
c     Description:Exercise the FORTRAN interface
c
c
      subroutine starttest(number)
      implicit none
      integer*4 number
      integer*8 tracker, throwaway
      call synch_RegOut_getInstance_f(tracker, throwaway)
      call synch_RegOut_startPart_f(tracker, number, throwaway)
      call synch_RegOut_deleteRef_f(tracker, throwaway)
      end

      subroutine reporttest(test, number,  python)
      implicit none
      integer*4 number
      integer*8 tracker, throwaway
      logical test, python
      include 'synch_ResultType.inc'
      call synch_RegOut_getInstance_f(tracker, throwaway)
      if (test) then
         call synch_RegOut_endPart_f(tracker, number, PASS,
     $        throwaway)
      else
         if (python) then
            call synch_RegOut_endPart_f(tracker, number, XFAIL,
     $           throwaway)
         else
            call synch_RegOut_endPart_f(tracker, number, FAIL,
     $           throwaway)
         endif
      endif
      call synch_RegOut_deleteRef_f(tracker, throwaway)
      number = number + 1
      end

      subroutine catch(exc)
      implicit none
      integer*8 exc, classinfo, tracker, throwaway
      character*(1024) buffer, name
      if (exc .ne. 0) then
         call synch_RegOut_getInstance_f(tracker, throwaway)
         call synch_RegOut_writeComment_f(tracker,
     $        'Unexpected exception thrown', throwaway)
         call sidl_BaseInterface_getClassInfo_f(exc, classinfo,
     $        throwaway)
         if (classinfo .ne. 0) then
            call sidl_ClassInfo_getName_f(classinfo, name, throwaway)
            buffer = 'Exception name: ' // name
            call synch_RegOut_writeComment_f(tracker, buffer,
     $           throwaway)
         end if
         call synch_RegOut_forceFailure_f(tracker, throwaway)
         call synch_RegOut_deleteRef_f(tracker, throwaway)
      end if
      exc = 1000
      end
      
      subroutine makeObject(obj, remoteURL)
      integer*8 obj
      character*256 remoteURL
      integer*8 throwaway

      if ( remoteURL .ne. '' ) then
         call hooks_Basics__createRemote_f(obj,remoteURL,throwaway)
      else
         call hooks_Basics__create_f(obj,throwaway)
      endif
      
      call catch(throwaway)
      end
      
      subroutine testhooks(test, remoteURL)
      implicit none
      integer*8 obj
      integer*8 throwaway
      integer*4 test, a, b, c, ret
      logical   boolval
      character*256 remoteURL
      
      boolval = .true.
      call hooks_Basics__set_hooks_static_f(boolval,throwaway)

      call makeObject(obj, remoteURL)
      call hooks_Basics__set_hooks_f(obj,boolval,throwaway)

      test = 1
      b = -1
      c = -1
      ret = 0

      call starttest(test)
      call hooks_Basics_aStaticMeth_f(test,b,c,ret,throwaway)
      call reporttest(b .eq. 1 .and. c .eq. 0, test, .false.)

      call starttest(test)
      call hooks_Basics_aStaticMeth_f(test,b,c,ret,throwaway)
      call reporttest(b .eq. 2 .and. c .eq. 1, test, .false.)

      b = -1
      c = -1
      ret = 0

      call starttest(test)
      call hooks_Basics_aNonStaticMeth_f(obj, test,b,c,ret,throwaway)
      call reporttest(b .eq. 1 .and. c .eq. 0, test, .false.)

      call starttest(test)
      call hooks_Basics_aNonStaticMeth_f(obj, test,b,c,ret,throwaway)
      call reporttest(b .eq. 2 .and. c .eq. 1, test, .false.)

      call hooks_Basics_deleteRef_f(obj,throwaway)
      end


      program hookstest
      integer*4 test, i
      integer*8 tracker, throwaway
      character*80 language
      character*256 remoteURL, arg
      logical result

      throwaway = 0
      call synch_RegOut_getInstance_f(tracker, throwaway)
      
C     Parse the command line  to see if we are running RMI tests
      remoteURL = ''
      do 10 i = 1, IArgc()
         call GetArg(i, arg)
         if ( arg(1:6) .eq. '--url=' ) then
            remoteURL = arg(7:)
         endif
 10   continue

C     Setup RMI if necessary
      if ( remoteURL .ne. '' ) then
         call GetArg(0, arg)
         call synch_RegOut_replaceMagicVars_f(tracker, remoteURL, arg,
     $        throwaway)
         write(*,'(A,A)') 'using remote URL ', remoteURL
         write(*,'(A)') 'registering RMI protocol simhandle'
         call sidl_rmi_ProtocolFactory_addProtocol_f('simhandle',
     $        'sidlx.rmi.SimHandle', result, throwaway)
         if (result .neqv. .TRUE. ) then
            write(*,'(A)') 'sidl.rmi.ProtocolFactor.addProtocol()
     $       failed'
         endif
      endif

      language = ' '
      call synch_RegOut_setExpectations_f(tracker, 4, throwaway)
      call synch_RegOut_writeComment_f(tracker,
     $     'Make sure hooks work', throwaway)
      call testhooks(test, remoteURL)
      call synch_RegOut_close_f(tracker, throwaway)
      call synch_RegOut_deleteRef_f(tracker,throwaway)
      end

File: regression/hooks/runC/hookstest.c
/*
 * File:        hookstest.c
 * Copyright:   (c) 2004 Lawrence Livermore National Security, LLC
 * Revision:    @(#) $Revision: 4533 $
 * Date:        $Date: 2005-04-13 16:38:20 -0700 (Wed, 13 Apr 2005) $
 * Description: Assertion Test C client
 *
 */
#include <stdio.h>
#include "hooks_Basics.h"
#include "sidl_Exception.h"
#include <stdlib.h>
#include <string.h>
#include "synch.h"
#include "sidl_Exception.h"
#include "sidl_rmi_ProtocolFactory.h"

/* #define RMI 0 */

#ifndef FALSE
#define FALSE 0
#endif
#ifndef TRUE
#define TRUE 1
#endif

#define MYASSERT( AAA ) \
  declare_part( tracker, &part_no ); \
  magicNumber = clearstack(magicNumber); \
  synch_RegOut_writeComment(tracker, #AAA, &exception); SIDL_REPORT(exception); \
  { \
     int _result = AAA; SIDL_REPORT(exception); \
     if ( _result ) result = synch_ResultType_PASS; \
     else result = synch_ResultType_FAIL;  \
  } \
  end_part( tracker, part_no, result);


void declare_part( synch_RegOut tracker, int *part_no );
void end_part( synch_RegOut tracker, int part_no, 
               enum synch_ResultType__enum result);

/**
 * Fill the stack with random junk.
 */
int clearstack(int magicNumber) {
  int chunk[2048], i;
  for(i = 0; i < 2048; i++){
    chunk[i] = rand() + magicNumber;
  }
  for(i = 0; i < 16; i++){
    magicNumber += chunk[rand() & 2047];
  }
  return magicNumber;
}

void declare_part( synch_RegOut tracker, int * part_no ) {
  sidl_BaseInterface exception = NULL;
  synch_RegOut_startPart(tracker, ++(*part_no), &exception);
  if (exception) {
    sidl_BaseInterface throwaway_exception;
    sidl_BaseInterface exception2 = NULL;
    synch_RegOut_forceFailure(tracker, &exception2);
    if (exception2) {
      puts("TEST_RESULT FAIL\n");
      sidl_BaseInterface_deleteRef(exception2, &throwaway_exception);
    }
    sidl_BaseInterface_deleteRef(exception, &throwaway_exception);
  }
}

void end_part( synch_RegOut tracker, int part_no, 
               enum synch_ResultType__enum result) {
  sidl_BaseInterface exception = NULL;
  synch_RegOut_endPart(tracker, part_no, result, &exception);
  if (exception) {
    sidl_BaseInterface throwaway_exception;
    sidl_BaseInterface exception2 = NULL;
    synch_RegOut_forceFailure(tracker, &exception2);
    if (exception2) {
      puts("TEST_RESULT FAIL\n");
      sidl_BaseInterface_deleteRef(exception2, &throwaway_exception);
    }
    sidl_BaseInterface_deleteRef(exception, &throwaway_exception);
  }
}

/* basic RMI support */
static char *remoteURL = NULL;

static sidl_bool withRMI( void ) {
  return remoteURL && *remoteURL;
}

static hooks_Basics
makeObject( void ) {
  hooks_Basics obj = NULL;
  sidl_BaseInterface exception = NULL;

#ifdef WITH_RMI  
  if(withRMI()) {
    obj = hooks_Basics__createRemote(remoteURL, &exception); SIDL_REPORT(exception);
    return obj;
  }
#endif
  
  obj = hooks_Basics__create(&exception); SIDL_REPORT(exception);
  return obj;
  EXIT:
  exit(1);
}

/*************************** main ***************************/
int main(int argc, char**argv)
{ 
  hooks_Basics h;
  sidl_BaseInterface exception = NULL;
  int32_t b, c, ret = 0;
  synch_RegOut tracker = synch_RegOut_getInstance(&exception);
  enum synch_ResultType__enum result = synch_ResultType_PASS;
  int part_no = 0;
  int magicNumber = 1;

  /* Parse the command line  to see if we are running RMI tests */
#ifdef WITH_RMI  
  {
    unsigned _arg;
    for(_arg = 0; _arg < argc; ++_arg) {
      if(strncmp(argv[_arg], "--url=", 6) == 0) {
        remoteURL = argv[_arg] + 6;
        synch_RegOut_replaceMagicVars(tracker, &remoteURL, argv[0], &exception); SIDL_REPORT(exception);        
      }
    }
    if(withRMI()) {
      fprintf(stdout, "using remote URL %s\n", remoteURL);
    }
  }

  /* Setup RMI if necessary */
  if(withRMI()) {
    sidl_BaseInterface _ex = NULL;
    fprintf(stdout, "registering RMI protocol simhandle\n");
    if(!sidl_rmi_ProtocolFactory_addProtocol("simhandle","sidlx.rmi.SimHandle",&_ex )){
      printf("sidl.rmi.ProtocolFactor.addProtocol() failed\n");
      exit(2);
    }
  }
#endif
  
  synch_RegOut_setExpectations(tracker, 4, &exception); SIDL_REPORT(exception);

  hooks_Basics__set_hooks_static(TRUE, &exception);SIDL_REPORT(exception);
  h = makeObject();
  hooks_Basics__set_hooks(h, TRUE, &exception);SIDL_REPORT(exception);

  b = c = -1;
  ret = hooks_Basics_aStaticMeth(part_no, &b, &c, &exception);SIDL_REPORT(exception);
  MYASSERT( b == 1 && c == 0 );
  ret = hooks_Basics_aStaticMeth(part_no, &b, &c, &exception);SIDL_REPORT(exception);
  MYASSERT( b == 2 && c == 1 );
  
  b = c = -1;
  ret = hooks_Basics_aNonStaticMeth(h, part_no, &b, &c, &exception);SIDL_REPORT(exception);
  MYASSERT( b == 1 && c == 0 );
  ret = hooks_Basics_aNonStaticMeth(h, part_no, &b, &c, &exception);SIDL_REPORT(exception);
  MYASSERT( b == 2 && c == 1 );

  hooks_Basics_deleteRef(h, &exception);SIDL_REPORT(exception);
  synch_RegOut_close(tracker, &exception);SIDL_REPORT(exception);
  synch_RegOut_deleteRef(tracker,&exception);SIDL_REPORT(exception);
  return 0;
 EXIT:
  {
    sidl_BaseInterface throwaway_exception = NULL;
    if (tracker) {
      sidl_BaseInterface exception2 = NULL;
      synch_RegOut_forceFailure(tracker, &exception2);
      if (exception2) {
	puts("TEST_RESULT FAIL\n");
	sidl_BaseInterface_deleteRef(exception2, &throwaway_exception);
      }
      synch_RegOut_deleteRef(tracker, &throwaway_exception);
    }
    sidl_BaseInterface_deleteRef(exception, &throwaway_exception);
    return -1;
  }
  return -1;
}
